.github/workflows/repo-check.yml

```diff
@@ -15,10 +15,15 @@ permissions:
   actions: write
 
 on:
+  pull_request: []
   push:
+    branches:
+      - master
+      - next
```

Comment:
Maybe I am missing something, but wouldn't this make `repo-check` run essentially like the other two checks?

This was the original state of this check, before we decided that we'd like to run each check as soon as possible in the development cycle, so problems are caught and resolved sooner rather than later. (e.g. we wanted to warn early if someone accidently added a large file to the repo during development).

Let's say we remove the `branches` specification so the triggers are just `pull_request:` and `push:`. (We can also check the `github.repository` so we don't run inside forks.)

This will be OK, except we will run twice for pushes to PRs from branches inside the mfem repo. Do you know if there is a way to check if:

- We are executing the `push` or `pull_request` trigger?

- The branch we are pushing to has an open PR?

Reply:
>This was the original state of this check, before we decided that we'd like to run each check as soon as possible in the development cycle, so problems are caught and resolved sooner rather than later. (e.g. we wanted to warn early if someone accidently added a large file to the repo during development).

That makes sense, but it comes at the cost that the CI is confusing. It also assumes the average contributor is going to check their own action runs (unlikely I think?) - and personally speaking, I don't check my fork CI before opening a PR. The intent of opening the PR is to start these checks. 

Maybe it would make sense to have this workflow run as a first job, and then cancel the remaining (more compute intensive) checks if those fail and send back a message to the user? That way you can get the quick feedback and have the checks run on the PR.

> Let's say we remove the branches specification so the triggers are just pull_request: and push:. (We can also check the github.repository so we don't run inside forks.)

That would reproduce the original problem because any push to the fork would trigger the CI there. That's what I was trying to fix. The github.repository check is also fairly useless (doesn't do anything) I was testing it out hoping that it all checks were cancelled it would be moved upstream to where we are checking the CI.

> We are executing the push or pull_request trigger?
>     The branch we are pushing to has an open PR?

Yes you can get this metadata alongside the action event. But I still think it doesn't make sense to have all these extra checks on pushes. Arguably any push should be reviewed via a PR, where it's checked there first! Generally the way I think about actions (if it helps):

- push events are for building production artifacts, containers, etc.
- pull requests do all the linting/ checking / testing

And if the developer needs to run tests you can use pre-commit or similar to make that easy.  Not sure if any of that helps, but it's the practices I've used and they work fairly well for issues like this!

Reply:
Okay, let me think about it a bit.

Reply:
Might I ask - what is the purpose of creating a fork and feature branch of you don’t intend to PR?

Reply:
Sorry, I meant branches above. Most of our PRs are from internal branches, not forks.

I can't speak for everyone, but there are many branches that are used for research of ideas that may or may not work, and so the goal is not to produce a merge-able code but to demo if something is possible. Some branches are used for collaboration, e.g. are associated with an LDRD project, and contain a lot of project-specific code that is not intended to be merged. Some branches could be temporary, e.g. a few branches with alternative solutions to a particular problem. Folks may also be reluctant to issue a PR before they feel the code is close to be ready. This is probably a big one, and probably a leftover of the days when there were no "Draft PRs". 

There really is a big variety and we intentionally left branch rules open and flexible in the past. Maybe we should rethink that. I will discuss it with the team this week.

Reply:
Oh gotcha! So if these branches need testing you could provide the same checks via something like pre-commit, or a workflow that just runs on push but has a pattern to skip all PRs. I’ve never done that before but if it’s flagged to skip PR I would hope it doesn’t show up in the CI there!

Reply:
So maybe I'd try like:

```yaml
on:

  # Run on every push of the branch
  push:

  # Never run or show up on any pull request!
  pull_request:
    branches-ignore:
      - '*'
```

Reply:
> something like pre-commit

We do have [that](https://github.com/mfem/mfem/tree/master/config/githooks), including `make hooks` for easy setup, and very much encourage people to use them

Reply:
Woot! Nice. 👍

Reply:
> So maybe I'd try like:

How will PRs be handled in the above proposal?

Reply:
The workflow above would not be for pull requests - it would be for this other use case of "branches that exist that we have no intention to PR."